# Falcon2Fastg 

This software converts the results of PacBio assembly using [FALCON](https://github.com/PacificBiosciences/FALCON), to a [FASTG](http://fastg.sourceforge.net/) graph that can be visualized using [Bandage](http://rrwick.github.io/Bandage/).

### Usage

    python Falcon2Fastg.py [--mode=read/contig/both; default: both]

Run it in the output directory of FALCON assembly (2-asm-falcon). It needs 
the following input files:

* preads4falcon.fasta

* sg_edges_list 

* ctg_paths (if `--mode` is set to `contig` or `both`)

* p_ctg.fa (if `--mode` is set to `contig` or `both`)  


### Dependencies :

Biopython (available at http://biopython.org/wiki/Download)

pyfaidx (available at https://github.com/mdshw5/pyfaidx)

Quick installation of dependencies:

    pip install biopython pyfaidx  # add --user if you don't have root

### Output : 

The output of the tool is two FASTG files (reads.fastg and contigs.fastg) that can be opened with 
[Bandage](http://rrwick.github.io/Bandage/).

![Alt text](/img/Falcon2Fastg_after_bandage.png?raw=true "Falcon2Fastg after Bandage")


Above is a sample Bandage visualization of a reads.fastg file generated by 
Falcon2Fastg from a FALCON assembly (plant mitochondrial genome).

* Each node is a read, and each node is represented as a colored strip 
(colors are random)
* Edges represent the overlaps between reads found by FALCON (better viewed in 
the zoomed-in image below) 
* Only the edges used in the string graph ("G" flagged in sg_edges_list) are 
used by Falcon2Fastg to produce the output file.



Zooming in on a smaller set of nodes shows the edges in black, connecting the colored nodes :


![Alt text](/img/zoomed_mito.png?raw=true "Falcon2Fastg zoom with Bandage")




For benchmarking, Falcon2Fastg was run on the preads4falcon.fasta and 
sg_edges_list file produced by the E.coli test dataset provided with the Falcon
 install. Instructions on obtaining the dataset are here : 
https://github.com/PacificBiosciences/FALCON/wiki/Setup:-Complete-example  

This run was performed on a desktop system with Intel Xeon CPU W3520 @ 2.67GHz 
and 4 cores. Runtime for this dataset (449 Mb) of preads4falcon.fasta is :

    time python Falcon2Fastg.py

    real	2m14.395s
    user	1m42.925s
    sys		0m9.514s


The figure below represents a visualization of this E. coli data.  

![Alt text](/img/ecoli_Gnodes.png?raw=true "Ecoli 'G' edges fastg after Bandage")


### Contig mode 

Falcon2Fastg can also be used to visualize overlaps between contigs. 

Usage : python Falcon2Fastg.py --mode=contig

This produces the output file "contigs.fastg"


To test this mode, we assembled Drosophila melanogaster reads available at 
(https://github.com/PacificBiosciences/DevNet/wiki/Drosophila-sequence-and-assembly). The FALCON assembly parameters were not optimized, and the 
final p_ctgs.fa file had 642 contigs with total length ~27 Mbp.  

This assembly was converted into a FASTG file using Falcon2Fastg --mode=contig

The run was performed on a desktop system with Intel Xeon CPU W3520 @ 2.67GHz
and 4 cores. Runtime for this dataset (2.1 Gb preads4falcon.fasta) is :

    time python Falcon2Fastg.py

    real    2m32.962s
    user    1m47.811s
    sys     0m10.254s


The figure below is the visualization of these D. mel. contigs.

![Alt text](/img/Dmel_ctgs.png?raw=true "D. mel. contigs after Bandage")



### Memory Warning 

The pyfaidx module is used to read an entire FASTA file into memory. 
If the size of your preads4falcon.fasta is greater than the amount of available
 RAM, it is advisable to run this computation on a server with greater
available memory. 



### Caveats : 

Faked a constant coverage of "50" because Bandage expects coverage information
 in the FASTG record headers.

Length record in FASTG header set to a constant "500". However, this does not 
seem to affect Bandage, which correctly calculates the read length


### Testing :

Please see the test/ directory for a small example dataset and output

FALCON can be installed following the instructions here : 
https://github.com/PacificBiosciences/FALCON/wiki/Setup:-Complete-example

## Other tools


The ```utils/graph_of_overlaps.sh``` script is included for advanced users only, to visualize a different type of graph: one where the nodes correspond to overlaps between reads (in the ```Falcon2Fastg.py```, overlaps are represented by edges, not nodes).

Usage:

    ./graph_of_overlaps.sh edges_list reads.fasta

Caveats :

Assumed that sequence entry in FASTG file represents overlap between two reads.

Faked a constant coverage of "50" because Bandage expects coverage information 
in the FASTG record headers.

Assumed the "length" record in FASTG header refers to length of overlap. 



## License 

This content is released under MIT License(https://opensource.org/licenses/MIT)
Please see LICENSE.md for details.


## Authors

Primary author : Samarth Rangavittal, The Pennsylvania State University
                 (szr165@psu.edu)
                
Rayan Chikhi, University of Lille 1   

Jean-Stéphane Varré, University of Lille 1

