# Falcon2Fastg 

This software converts the results of PacBio assembly using [FALCON](https://github.com/PacificBiosciences/FALCON), to a [FASTG](http://fastg.sourceforge.net/) graph that can be visualized using [Bandage](http://rrwick.github.io/Bandage/).

### Usage

    python Falcon2Fastg.py

Run it in the output directory of FALCON assembly (2-asm-falcon). It needs 
the following input files:

* preads4falcon.fasta

* sg_edges_list 


### Dependencies :

Biopython (available at http://biopython.org/wiki/Download)


### Output : 

The output of the tool is a FASTG file (output.fastg) that can be opened with 
[Bandage](http://rrwick.github.io/Bandage/).

![Alt text](/img/Falcon2Fastg_after_bandage.png?raw=true "Falcon2Fastg after Bandage")


Above is a sample Bandage visualization of a FASTG file generated by 
Falcon2Fastg from a FALCON assembly (mitochondrial genome).

* Each node is a read, and each node is represented as a colored strip 
(colors are random)
* Edges represent the overlaps between reads found by FALCON (better viewed in 
the zoomed-in image below) 
* Only the edges used in the string graph ("G" flagged in sg_edges_list) are 
used by Falcon2Fastg to produce the output file.



Zooming in on a smaller set of nodes shows the edges in black, connecting the colored nodes :


![Alt text](/img/zoomed_mito.png?raw=true "Falcon2Fastg zoom with Bandage")




For benchmarking, Falcon2Fastg was run on the preads4falcon.fasta and 
sg_edges_list file produced by the E.coli test dataset provided with the Falcon
 install. Instructions on obtaining the dataset are here : 
https://github.com/PacificBiosciences/FALCON/wiki/Setup:-Complete-example  

This run was performed on a desktop system with Intel Xeon CPU W3520 @ 2.67GHz 
and 4 cores. Runtime for this dataset (449 Mb) of preads4falcon.fasta is :

    time python Falcon2Fastg.py

    real	4m8.053s
    user	2m13.920s
    sys		1m53.490s


The figure below represents a visualization of this E. coli data.  

![Alt text](/img/ecoli_Gnodes.png?raw=true "Ecoli 'G' edges fastg after Bandage")





### Caveats : 

Faked a constant coverage of "50" because Bandage expects coverage information
 in the FASTG record headers.

Length record in FASTG header set to a constant "500". However, this does not 
seem to affect Bandage, which correctly calculates the read length


### Testing :

Please see the test/ directory for a small example dataset and output

Falcon can be installed following the instructions here : 
https://github.com/PacificBiosciences/FALCON/wiki/Setup:-Complete-example

## Other tools


The ```utils/graph_of_overlaps.sh``` script is included for advanced users only, to visualize a different type of graph: one where the nodes correspond to overlaps between reads (in the ```Falcon2Fastg.py```, overlaps are represented by edges, not nodes).

Usage:

    ./graph_of_overlaps.sh edges_list reads.fasta

Caveats :

Assumed that sequence entry in FASTG file represents overlap between two reads.

Faked a constant coverage of "50" because Bandage expects coverage information 
in the FASTG record headers.

Assumed the "length" record in FASTG header refers to length of overlap. 



## License 

This content is released under MIT License(https://opensource.org/licenses/MIT)
Please see LICENSE.md for details.


## Authors

Primary author : Samarth Rangavittal, The Pennsylvania State University
                 (szr165@psu.edu)
  
Collaborators :  Rayan Chikhi, University of Lille 1                 
                 Jean-Stéphane Varré, University of Lille 1

